### PR TITLE
Fix excess property checking for nullary sums

### LINKED ISSUE #3: No excess property checks for nullary sums
The following produces a runtime exception with no type errors.

```ts
import { getShow } from '@unsplash/sum-types-fp-ts';
import * as Sum from '@unsplash/sum-types';

type S = Sum.Member<'A'> | Sum.Member<'B'>;
const S = Sum.create<S>();

const show = getShow<S>({
  A: 'Custom',
});

console.log(show.show(S.mk.A));
```

https://stackblitz.com/edit/typescript-5afkt4?file=index.ts

I believe if supplied, `A` must be an instance of `Show` but TS doesn't check for it. 

Side note, I find it a little confusing prepending the sum member key in the show output here https://github.com/unsplash/sum-types-fp-ts/blob/master/src/index.ts#L153 . In the following code I'm expecting the show output to be `CustomA` but instead I get `A CustomerA`

```ts
const show = getShow<S>({
  A: {
    show: () => "CustomA"
  }  
});

```

### PR DESCRIPTION
**Pull request by samhh**
Fixes #3 with only type-level changes.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -18,18 +18,36 @@ type Value<A> = A extends Sum.Member<any, infer B> ? B : never
...
-  readonly [B in A as Value<B> extends null ? never : Tag<B>]: Show<Value<B>>
-}
+//
+// If a sum is nullary then the reduced input type would be `{}`, which
+// unfortunately disables excess property checking, hence the conditional. We
+// wrap the condition in tuples to avoid distributing over the union.
+//
+// It should be noted that excess property checking won't help us if the input
+// object is first defined elsewhere and then provided to our function. This
+// should be considered an unsafe, unsupported use of this API.
```
Comment (astlouisf) [inline src/index.ts:28]: Wasn't this already a limitation?

File: src/index.ts [hunk 2]
```diff
@@ -66,8 +84,11 @@ export const getEq = <A extends Sum.AnyMember>(eqs: Eqs<A>): Eq<A> =>
 
     const eq = eqs[xk as keyof typeof eqs]
 
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    return eq === undefined || eq.equals(xv as any, yv as any)
+    return (
+      eq === undefined ||
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (eq as unknown as Eq<Value<A>>).equals(xv as any, yv as any)
```
Comment (astlouisf) [inline src/index.ts:90]: Do you know how `eq` gets its type `const eq: Eqs<A>[keyof Eqs<A>] & ({} | null)` after the equality test with undefined while it was simply `const eq: Eqs<A>[keyof Eqs<A>]` before it?

File: src/index.ts (hunk 1 above)
Comment (samhh) [reply src/index.ts:28]: Correct, it just seemed worth writing down.

File: src/index.ts (hunk 2 above)
Comment (samhh) [reply src/index.ts:90]: I'm not seeing that inferred intersection. I wouldn't waste much time on it though, these libraries are all absolutely littered with assertions. I don't think there's much we can do about it (though you're welcome to try 😉). The only real concern for me is what's consumer-facing.

**Pull request merged**